cpp/include/raft/distance/detail/masked_nn.cuh

```diff
@@ -0,0 +1,308 @@
+                  CoreLambda,
+                  decltype(epilog_lambda),
+                  FinalLambda,
+                  decltype(rowEpilog_lambda),
+                  true>
+    obj(x,
+        y,
+        m,
+        n,
+        k,
+        lda,
+        ldb,
+        ldd,
+        xn,
+        yn,
+        adj,
+        group_idxs,
+        num_groups,
+        smem,
+        core_op,
+        epilog_lambda,
+        fin_op,
+        rowEpilog_lambda);
+  obj.run();
+}
+
+template <typename DataT, typename OutT, typename IdxT, typename ReduceOpT, typename KVPReduceOpT>
+void maskedL2NNImpl(raft::handle_t& handle,
+                    OutT* min,
+                    const DataT* x,
+                    const DataT* y,
+                    const DataT* xn,
+                    const DataT* yn,
+                    const bool* adj,
+                    const IdxT* group_idxs,
+                    IdxT num_groups,
+                    IdxT m,
+                    IdxT n,
+                    IdxT k,
+                    ReduceOpT redOp,
+                    KVPReduceOpT pairRedOp,
+                    bool sqrt,
+                    bool initOutBuffer)
+{
+  typedef typename linalg::Policy4x4<DataT, 1>::Policy P;
+
+  static_assert(P::Mblk == 64, "maskedL2NNImpl only supports a policy with 64 rows per block.");
+
+  // Get stream and workspace memory resource
+  rmm::mr::device_memory_resource* ws_mr =
+    dynamic_cast<rmm::mr::device_memory_resource*>(handle.get_workspace_resource());
+  auto stream = handle.get_stream();
+
+  // Acquire temporary buffers and initialize to zero:
+  // 1) Adjacency matrix bitfield
+  // 2) Workspace for fused nearest neighbor operation
+  size_t m_div_64 = raft::ceildiv(m, IdxT(64));
+  rmm::device_uvector<uint64_t> ws_adj64{m_div_64 * num_groups, stream, ws_mr};
+  rmm::device_uvector<int> ws_fused_nn{size_t(m), stream, ws_mr};
+  RAFT_CUDA_TRY(cudaMemsetAsync(ws_adj64.data(), 0, ws_adj64.size() * sizeof(uint64_t), stream));
+  RAFT_CUDA_TRY(cudaMemsetAsync(ws_fused_nn.data(), 0, ws_fused_nn.size() * sizeof(int), stream));
+
+  // Compress boolean adjacency matrix to bitfield.
+  dim3 compress_grid(raft::ceildiv(m, 32), raft::ceildiv(num_groups, 32));
+  compress_to_bits_naive<<<compress_grid, dim3(32, 32), 0, stream>>>(
+    adj, num_groups, m, ws_adj64.data());
+
+  dim3 blk(P::Nthreads);
+  auto nblks            = raft::ceildiv<int>(m, P::Nthreads);
+  constexpr auto maxVal = std::numeric_limits<DataT>::max();
+  typedef raft::KeyValuePair<IdxT, DataT> KVPair;
+
+  // Accumulation operation lambda
+  auto core_lambda = [] __device__(DataT & acc, DataT & x, DataT & y) { acc += x * y; };
+
+  if (initOutBuffer) {
+    initKernel<DataT, OutT, IdxT, ReduceOpT>
+      <<<nblks, P::Nthreads, 0, stream>>>(min, m, maxVal, redOp);
+    RAFT_CUDA_TRY(cudaGetLastError());
+  }
+
+  auto fin_op = raft::identity_op{};
+
+  constexpr size_t shmemSize = P::SmemSize + ((P::Mblk + P::Nblk) * sizeof(DataT));
```

Comment:
Right now, it could. Since the function only supports using the L2 metric. However, for other metrics such as the the L1 metric that do not require storing the norms in shared memory, the  `((P::Mblk + P::Nblk) * sizeof(DataT))` could be removed. In addition, there is the correlation metric that requires two norms to be stored. 

All in all, I do agree that we could benefit from more abstraction and separation of concerns, but I think we should leave this task to PR #1142 . I have added a note there.